commit 3b1b3eaa36a409377fc34fa6497526b51a5d378e (refs/changes/94/394/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-09-01T14:24:37-07:00 (3 years, 1 month ago)
    
    joyent/node-cueball#15 Backends appear without IP addresses when using CNAMEs

diff --git a/lib/resolver.js b/lib/resolver.js
index 714bea4..ebc004d 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -774,12 +774,21 @@ CueBallDNSResolver.prototype.resolve = function (domain, type, timeout) {
 		var answers = msg.getAnswers();
 		var minTTL = undefined;
 		if (type === 'A' || type === 'AAAA') {
-			ans = answers.map(function (a) {
+			ans = [];
+			answers.forEach(function (a) {
+				if (a.type !== type) {
+					if (a.type === 'CNAME' ||
+					    a.type === 'DNAME')
+						return;
+					self.r_log.warn('got unsupported ' +
+					    'answer rrtype: %s', a.type);
+					return;
+				}
 				if (minTTL === undefined ||
 				    a.ttl < minTTL) {
 					minTTL = a.ttl;
 				}
-				return ({
+				ans.push({
 					name: a.name,
 					address: a.target
 				});
@@ -789,6 +798,14 @@ CueBallDNSResolver.prototype.resolve = function (domain, type, timeout) {
 			var cache = {};
 			var addns = msg.getAdditionals();
 			addns.forEach(function (rr) {
+				if (rr.type !== type) {
+					if (rr.type === 'CNAME' ||
+					    rr.type === 'DNAME')
+						return;
+					self.r_log.warn('got unsupported ' +
+					    'additional rrtype: %s', rr.type);
+					return;
+				}
 				if (rr.target) {
 					if (minTTL === undefined ||
 					    rr.ttl < minTTL) {
@@ -799,7 +816,16 @@ CueBallDNSResolver.prototype.resolve = function (domain, type, timeout) {
 					cache[rr.name].push(rr.target);
 				}
 			});
-			ans = answers.map(function (a) {
+			ans = [];
+			answers.forEach(function (a) {
+				if (a.type !== type) {
+					if (a.type === 'CNAME' ||
+					    a.type === 'DNAME')
+						return;
+					self.r_log.warn('got unsupported ' +
+					    'answer rrtype: %s', a.type);
+					return;
+				}
 				if (minTTL === undefined ||
 				    a.ttl < minTTL) {
 					minTTL = a.ttl;
@@ -807,7 +833,7 @@ CueBallDNSResolver.prototype.resolve = function (domain, type, timeout) {
 				var obj = { name: a.target, port: a.port };
 				if (cache[a.target])
 					obj.additionals = cache[a.target];
-				return (obj);
+				ans.push(obj);
 			});
 
 		} else {
